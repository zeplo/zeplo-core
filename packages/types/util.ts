
export type Maybe<T> = T | null

export type JsonValue =  boolean | number | string | null | JsonArray | JsonMap
export interface JsonMap { [key: string]: JsonValue }
export interface JsonArray extends Array<JsonValue> {}

export type DeepPartial<T> = T extends Function ? T : (T extends object ? { [P in keyof T]?: DeepPartial<T[P]>; } : T);

/**
 * Get all of the keys to which U can be assigned.
 */
export type OnlyKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? K : never
}[keyof T];

/**
 * Get the interface containing only properties to which U can be assigned.
 */
export type OnlyUndefined<T> = {
  [K in OnlyKeys<T, undefined>]: T[K]
}

/**
 * Get all of the keys except those to which U can be assigned.
 */
export type ExcludeKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? never : K
}[keyof T];

/**
 * Get the interface containing no properties to which U can be assigned.
 */
export type ExcludeUndefined<T> = {
  [K in ExcludeKeys<T, undefined>]: T[K]
}

/**
 * Get the interface where all properties are optional.
 */
export type Optional<T> = {[K in keyof T]?: T[K]};

/**
 * Get the interface where properties that can be assigned undefined are
 * also optional.
 */
export type UndefinedOptional<T> = ExcludeUndefined<T> & Optional<OnlyUndefined<T>>

export type DeepUndefinedOptional<T> = T extends (...args: any[]) => any
  ? T
  : T extends any[]
  ? _DeepRequiredArray<T[number]>
  : T extends object
  ? _DeepRequiredObject<UndefinedOptional<T>>
  : T;

 export interface _DeepRequiredArray<T>
  extends Array<DeepUndefinedOptional<T>> {}

  export type _DeepRequiredObject<T> = {
  [P in keyof T]: DeepUndefinedOptional<T[P]>
}