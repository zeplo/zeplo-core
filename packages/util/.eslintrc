{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "parser":  "@typescript-eslint/parser",
  "env": {
    "browser": true,
    "node": true,
    "es6": true,
    "mocha": true,
    "jest": true
  },
  "rules": {
    "semi": ["error", "never"],
    "no-underscore-dangle": 0,
    "space-before-function-paren": ["error", "always"],
    "arrow-body-style": 0,
    "no-use-before-define": ["error", { "functions": false }],
    "arrow-parens": 0,
  }
}
