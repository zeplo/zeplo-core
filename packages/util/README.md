# @zeplo/util

Utility functions that are used across packages.