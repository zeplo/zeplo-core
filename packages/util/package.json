{
  "name": "@zeplo/util",
  "version": "2.1.0",
  "main": "dist/index.js",
  "files": [
    "dist/*"
  ],
  "license": "MIT",
  "scripts": {
    "clean": "rimraf dist",
    "dev": "env-cmd ts-node",
    "build": "tsc",
    "prepublish": "yarn clean && yarn build"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/fs-extra": "^9.0.13",
    "@types/lodash": "^4.14.178",
    "@types/node": "^17.0.18",
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "env-cmd": "^10.1.0",
    "eslint": "^8.9.0",
    "rimraf": "^3.0.2",
    "ts-node": "^10.5.0",
    "ts-node-dev": "^1.1.8",
    "typescript": "^4.5.5"
  },
  "dependencies": {
    "@zeplo/errors": "^2.0.8",
    "@zeplo/types": "^2.1.0",
    "express": "^4.17.3",
    "express-async-handler": "^1.2.0",
    "lodash": "^4.17.21",
    "uuid": "^8.3.2"
  }
}
