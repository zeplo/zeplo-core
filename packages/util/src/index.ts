export { cleanQuery, cleanHeaders } from './clean'
export { 
  parseRequest, 
  parseBody, 
  getBackoff, 
  getStartTimeFromFeatures,
  getRawFeatures,
  parseRawRequestObject, 
  createRequestJob ,
  getValueFromQueryOrHeader,
} from './parse'
export { 
  createStepsJobs,
  hasCyclicalDependency,
} from './steps'
export { 
  requestHandler,
  jsonRequestHandler,
  queryToString,
  queryToNumber,
  parseQuery,
} from './request'