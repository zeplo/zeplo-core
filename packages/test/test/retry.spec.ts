/**
 * @jest-environment node
 */

import { omit } from 'lodash'
import { createWorkspace } from '#/util/data'
import waitFor from '#/util/waitFor'
import wait from '#/util/wait'
import queue from '#/util/queue'

jest.setTimeout(60000)

describe('retry.spec', () => {
  const workspace = `test-workspace-retry-${Date.now()}`
  const token = `test-token-retry-${Date.now()}`

  let cleanup: () => Promise<void>
  beforeAll(async () => {
    cleanup = await createWorkspace(workspace, token)
  })
  afterAll(() => cleanup())

  test('retry GET request * 3 with initial success - using _retry param', async () => {
    const res = await queue.get(`/https://postman-echo.com/get?_retry=${3}&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await queue.get(`/requests/${id}?_token=${token}`)

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      retry: { max: 3, backoff: 'FIXED', time: 1 },
      request: { method: 'GET' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('retry GET request with FIXED 8 second backoff', async () => {
    const res = await queue.post(`/https://postman-echo.com/get?_retry=1|fixed|8&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    await wait(4000)

    const before = await queue.get(`/requests/${id}?_token=${token}`)
    expect(before.data.attempts).toBe(1)

    await wait(6000)

    const after = await queue.get(`/requests/${id}?_token=${token}`)
    expect(before.data.attempts).toBe(1)

    expect(after.data).toMatchObject({
      attempts: 2,
      source: 'REQUEST',
      retry: { max: 1, backoff: 'FIXED', time: 8 },
    })
  })

  test('retry GET request with EXPONENTIAL 2 second backoff', async () => {
    const res = await queue.post(`/https://postman-echo.com/get?_retry=2|exponential|2&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    await wait(3000)

    const before = await queue.get(`/requests/${id}?_token=${token}`)
    expect(before.data.attempts).toBe(2)

    await wait(5000)

    const after = await queue.get(`/requests/${id}?_token=${token}`)
    expect(after.data.attempts).toBe(3)

    expect(after.data).toMatchObject({
      attempts: 3,
      source: 'REQUEST',
      retry: { max: 2, backoff: 'EXPONENTIAL', time: 2 },
    })
  })

  test('retry GET request with IMMEDIATE', async () => {
    const res = await queue.post(`/https://postman-echo.com/get?_retry=1|IMMEDIATE&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    await wait(3000)

    const req = await queue.get(`/requests/${id}?_token=${token}`)

    expect(req.data).toMatchObject({
      attempts: 2,
      source: 'REQUEST',
      retry: { max: 1, backoff: 'IMMEDIATE' },
    })
  })

  test('retry GET request * 3 with initial success - using X-Zeplo-Retry param', async () => {
    const res = await queue.get(`/https://postman-echo.com/get?_token=${token}`, {
      headers: { 'X-Zeplo-Retry': 3 },
    })
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await queue.get(`/requests/${id}?_token=${token}`)

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      retry: { max: 3, backoff: 'FIXED', time: 1 },
      request: { method: 'GET' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('retry GET request * 3 with 404 error', async () => {
    const res = await queue.get(`/https://postman-echo.com/post?_retry=${3}&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    await wait(4000)

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.attempts >= 4,
    )

    expect(req.data).toMatchObject({
      status: 'ERROR',
      source: 'REQUEST',
      attempts: 4,
      retry: { max: 3, backoff: 'FIXED', time: 1 },
      request: { method: 'GET' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const children = await queue.get(`/requests?_token=${token}&filters={"trace": "${id}" }`)
    expect(children.data.length).toBe(3)
    expect(children.data[0]).toMatchObject({
      status: 'ERROR',
      trace: id,
    })
  })

  test('retry with invalid param should default to 1 - AZ chars', async () => {
    const res = await queue.get(`/https://postman-echo.com/post?_retry=abc123asd&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.attempts >= 2,
    )

    expect(req.data).toMatchObject({
      attempts: 2,
      status: 'ERROR',
      source: 'REQUEST',
      retry: { max: 1, backoff: 'FIXED', time: 1 },
    })
  })

  test('retry GET request with unknown backoff approrach uses default', async () => {
    const res = await queue.post(`/https://postman-echo.com/get?_retry=1|unknown&_token|asdasd&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await queue.get(`/requests/${id}?_token=${token}`)

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      retry: { max: 1, backoff: 'FIXED', time: 1 },
    })
  })

  test('retry GET request with lowercase backoff', async () => {
    const res = await queue.post(`/https://postman-echo.com/get?_retry=1|fixed|2&_token|asdasd&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await queue.get(`/requests/${id}?_token=${token}`)

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      retry: { max: 1, backoff: 'FIXED', time: 2 },
    })
  })

  test('retry with too many params', async () => {
    const res = await queue.get(`/https://postman-echo.com/post?_retry=1|FIXED|1|another&_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.attempts >= 2,
    )

    expect(req.data).toMatchObject({
      attempts: 2,
      status: 'ERROR',
      source: 'REQUEST',
      retry: { max: 1, backoff: 'FIXED', time: 1 },
    })
  })
})
