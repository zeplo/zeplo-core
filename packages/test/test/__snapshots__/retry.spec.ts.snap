// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retry.spec retry GET request * 3 with 404 error 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`retry.spec retry GET request * 3 with initial success - using _retry param 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;

exports[`retry.spec retry GET request * 3 with initial success - using X-Zeplo-Retry param 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;
